### PR TITLE
Kokkos complex_align variant, Trilinos+PETSc enforcement for Kokkos~complex_align

### PR DESCRIPTION
**Pull request by finkandreas**
This PR tackles two problems, but I felt that I will add it to a single PR. Let me know if I must split it into two.
The first issue is the missing variant of `complex_align` on Kokkos. Trilinos, when built with the builtin kokkos version will set `complex_align` to OFF, which is the opposite of what the default in the Kokkos package is.
This PR adds the variant with default value true, but enforcing it to false, when building trilinos.
Also PETSc, when building for complex scalars and kokkos support enforces that the variant `complex_align` is set to false.

The second issue this PR tackles is the incorrect dependency on the external Kokkos package for trilinos. Basically this line
```
        if spec.satisfies("@14.4.0 +kokkos"):
```
This is missing the upper boundary, and only restricts on the very specific trilinos version.
This leads to trilinos pulling in the internal kokkos version, and one would end up with two kokkos libraries. One provided by kokkos (correct) and one provided by trilinos (incorrect).
Also I have the feeling that when we enable `trilinos+kokkos`, we actually also want to unbundle `kokkos-kernels` from trilinos, hence I added this too.
All in all, using this PR we can build a version of trilinos that links to `kokkos` and `kokkos-kernels` without pulling in the bundled versions inside trilinos.

Without this PR, building an environment with a view would lead to multiple `libkokkos_*.so`, provided by `trilinos`, `kokkos` and `kokkos-kernels`.

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Maybe slightly cleaner would be to extract the common when condition, and split the version pinning from general constraints to avoid repetition. Then you can maybe also comment why ~complex_align is necessary in a dedicated place

```python
with when("@14.4: +kokkos"):
    depends_on("kokkos~complex_align")  # generic requirement
    depends_on("kokkos@4.3.01", when="@16")  # bunch of version pinning statements
    ...
```

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/trilinos/package.py]: `@16` would [work too](https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-backward-and-forward-compatibility), maybe a bit nicer

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/trilinos/package.py:414]: should this then be `@17:` instead of `@master:` to cover the full range of possible versions?

Comment (finkandreas) [reply var/spack/repos/builtin/packages/trilinos/package.py:414]: Not sure, since 17 is not released, and probably will be a different version anyway. Depending on whatever trilinos will have pulled in by that time. I would leave this for the maintainer to add the correct version, once version 17 is added.

**Review by haampie — DISMISSED** [review]

**Comment by balay** [comment]
cc: @jczhang07

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
        depends_on("kokkos+wrapper", when="+wrapper")
        depends_on("kokkos~wrapper", when="~wrapper")
```

can this go before/after the ~complex_align statement?

Comment (haampie) [reply var/spack/repos/builtin/packages/trilinos/package.py]: maybe also the cuda_arch / amdgpu_arch bits with `.format(...)` replaced by more modern f-strings `f"+cuda cuda_arch={a}`... sorry for the many requests :)

Comment (finkandreas) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Addressed both.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/trilinos/package.py]: also one more level of indent and drop the `@14.4: +kokkos` bit in both places? could probably just fit in one line if `spack style` doesn't complain about lines getting too long

```python
depends_on(f"kokkos +cuda cuda_arch={a}", when=f"+cuda cuda_arch={a}")
```

this is the last request, I promise :p

Comment (finkandreas) [reply var/spack/repos/builtin/packages/trilinos/package.py]: Argh, I did not even check what this block was doing. 
Now I've shortened it vertically and moved to the common `with when` block.

**Review by haampie — APPROVED** [review]
Thanks, I'll let the trilinos/kokkos folks have a look, but in terms of style it's a big improvement as well.

**Comment by jczhang07** [comment]
> cc: @jczhang07

@balay yes, we added `-DKokkos_ENABLE_COMPLEX_ALIGN=OFF` in https://gitlab.com/petsc/petsc/-/merge_requests/5038.  I regret I did not log the error message in the MR. I guess it was because when building kokkos on CPUs, we want interoperability between std::complex and kokkos::complex.

**Comment by psakievich** [comment]
This seems fine to me, but I'm not a core member of the kokkos team.

**Review by cedricchevalier19 — APPROVED** [review]
Ok for the kokkos recipe.

**Review by psakievich — APPROVED** [review]

**Comment by balay** [comment]
This change is triggering `trilinos@14.4.0` build error with `./bin/spack install xsdk@1.0.0%gcc@11.4.0`

Before this change:
[spack-build-out-good-truncated.txt](https://github.com/user-attachments/files/17961563/spack-build-out-good-truncated.txt)

After this change:
[spack-build-out-bad.txt](https://github.com/user-attachments/files/17961568/spack-build-out-bad.txt)

**Pull request merged**
